sway/tree/container.c

```diff
@@ -378,3 +381,25 @@ void container_map(swayc_t *container, void (*f)(swayc_t *view, void *data), voi
 		f(container, data);
 	}
 }
+
+void container_for_each_bfs(swayc_t *con, void (*f)(swayc_t *con, void *data),
+		void *data) {
+	list_t *queue = create_list();
+	if (queue == NULL) {
+		wlr_log(L_ERROR, "could not allocate list");
+		return;
+	}
+
+	list_add(queue, con);
+
+	swayc_t *current = NULL;
+	while (queue->length) {
+		current = queue->items[0];
+		list_del(queue, 0);
+		f(current, data);
+		// TODO floating containers
+		list_cat(queue, current->children);
+	}
+
+	list_free(queue);
+}
```

Comment:
Just iterate over the children twice

```c
for child : con.children {
    f(child, data);
}
for child : con.children {
    container_for_each_bfs(child, f, data);
}
```

I still don't like allocating a list here

Reply:
I'm not sure how to do a bfs without allocating a queue. Maybe the queue could be allocated once and used for the lifetime of the application?

It gets worse when we want to do reverse bfs because that requires two lists: one for the queue, and one for the result in reverse order.

Event worse, I'm also thinking that BFS with containers in `children` array order might not truly be rendering order, but rather based on the focus-inactive stack. So a child later in the stack might need to be rendered on top even though they are siblings (if for instance it is the focused window). You can see this bug right now when the popups of the focused window will render below a sibling.

Reply:
Static queue sounds good to me